### PR TITLE
Initial support for proposed New wxStyledTextCtrl Autocompletion API

### PR DESCRIPTION
**Pull request by mtangoo**
I have added
- New `wxCompletionItem` class
- Added new Overload of `AutoCompShow(int len, const std::vector<wxCompletionItem> &list)`

Would like to know your opinion before I go further. If accepted I plan:
- [x] On Completed event, send the Client Object
- [x] Write Documentation including since when it was added
- [x] Write sample description of the feature

### CODE REVIEW

**Comment by PBfordev** [comment]
<I know nothing about Scintilla or wxSTC so I cannot comment on the actual PR, did not know from the PR title it is STC related but noticed somthing in the first message>

Just curious, was the "ban" on using the STL in wxWidgets code lifted (your code uses `std::vector` instead of the usual `wxVector`)? Scintilla already uses STL so perhaps there is an exception?

**Comment by mtangoo** [comment]
Hi @PBfordev ,
>I know nothing about Scintilla or wxSTC so I cannot comment on the actual PR, did not know from the PR title it is STC related but noticed somthing in the first message

Sorry for the confusion. I have edited the title. Thank you for pointing that out

>Just curious, was the "ban" on using the STL in wxWidgets code lifted (your code uses std::vector instead of the usual wxVector)? Scintilla already uses STL so perhaps there is an exception

Never knew of STL ban. @vadz can you shed light on the ban?

I'm used to STL in my code so It's natural that I didn't thought of wxVector. I will update it though!

**Comment by PBfordev** [comment]
> Never knew of STL ban. 

I meant [this](http://wxwidgets.org/develop/coding-guidelines/#no_stl). The rule probably dates back to when the (properly working) STL implentation was not available on all the platforms wxWidgets supported.

I believe that to this day, wxWidgets by default (see `wxUSE_STL` and `wxUSE_STD_*`) does not use any STL classes itself.

Sorry for PR hijacking.

**Comment by mtangoo** [comment]
>I meant this. The rule probably dates back to when the (properly working) STL implementation was not available on all the platforms wxWidgets supported. I believe that to this day, wxWidgets by default (see wxUSE_STL and wxUSE_STD_*) does not use any STL classes itself.

I had all but forgotten that page altogether. Thanks for the reminder. I have updated the code!

>Sorry for PR hijacking.

It wasn't hijacking at any rate. Very useful comments!
Cheers!

**Comment by oneeyeman1** [comment]
@mtangoo ,
The build chockes.
Did you forget to check in some files?

**Comment by mtangoo** [comment]
@oneeyeman1, I added build files only to CMake as that is what am testing on. Didn't bother anything else as the first version of the PR was to demo the addition and let the team here review and kind of giving green light. I'm trying to avoid putting effort on something that won't be accepted at the end of the day.

TL;DR: Ignore them for now

File: .gitignore [hunk 1]
```diff
@@ -475,3 +475,7 @@
 
 # /utils/wxrc/
 /utils/wxrc/wxrc_vc[789].sln
+
+#Ignore OS Specific files
+#MAC OS
+.DS_Store
```
Comment (vadz) [inline .gitignore]: Please don't do this in wx `.gitignore`, it should be done once for all projects in the global one and not here.

File: include/wx/stc/completionitem.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include "wx/string.h"
+#include "wx/clntdata.h"
+
+class WXDLLIMPEXP_STC wxCompletionItem
+{
+public:
+    wxCompletionItem(const wxString &name, int icon, const wxString &doc = wxString(), wxClientData *data = NULL);
+    virtual ~wxCompletionItem();
+
+    const wxString &GetName();
```
Comment (vadz) [inline include/wx/stc/completionitem.h]: This is a global style comment: please put `&` near the type, not the identifier. I.e.
```suggestion
    const wxString& GetName();
```

File: include/wx/stc/completionitem.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#define _WX_STC_COMPLETIONITEM_H_
+
+#include "wx/string.h"
+#include "wx/clntdata.h"
+
+class WXDLLIMPEXP_STC wxCompletionItem
+{
+public:
+    wxCompletionItem(const wxString &name, int icon, const wxString &doc = wxString(), wxClientData *data = NULL);
+    virtual ~wxCompletionItem();
```
Comment (vadz) [inline include/wx/stc/completionitem.h]: Is this class supposed to be polymorphic or not? If yes, it should have some virtual functions and it should _not_ use `wxClientData`, being a kind of client data on its own. If not, it shouldn't have a virtual dtor.

I'd prefer to make it polymorphic rather than using client data, which is a kludge used when there is no other possibility.

File: include/wx/stc/completionitem.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    const wxString &GetName();
+    int GetIcon();
+    const wxString &GetDocumentation();
+    wxClientData *GetClientObject();
+
+protected:
```
Comment (vadz) [inline include/wx/stc/completionitem.h]: Please use `private` unless you really want to allow derived class changing the members directly (and, if so, why?).

File: include/wx/stc/stc.h [hunk 5]
```diff
@@ -39,6 +39,9 @@
 #include "wx/stopwatch.h"
 #include "wx/versioninfo.h"
 
+#include "wx/vector.h"
+#include "completionitem.h"
```
Comment (vadz) [inline include/wx/stc/stc.h]:
```suggestion
#include "wx/stc/completionitem.h"
```
for consistency

File: include/wx/stc/stc.h [hunk 6]
```diff
@@ -5197,7 +5200,8 @@ class WXDLLIMPEXP_STC wxStyledTextCtrl : public wxControl,
                              const wxColour& currentBgColour = wxNullColour,
                              const wxColour& currentTextColour = wxNullColour);
 
-
+    //Show autocompletion with a more intelligent way
```
Comment (vadz) [inline include/wx/stc/stc.h]: This comment is not really helpful... I have no idea what does it really mean.

File: src/stc/completionitem.cpp [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#include "wx/stc/completionitem.h"
```
Comment (vadz) [inline src/stc/completionitem.cpp]: It's probably not necessary to have this file at all, considering how trivial it is you can just make all the methods inline.

But if you do add a new source file, please use the usual format for it, e.g. a header comment, PCH inclusion, etc.

File: src/stc/stc.cpp [hunk 8]
```diff
@@ -5062,7 +5062,19 @@ void wxStyledTextCtrl::AutoCompUseListCtrl(bool useListCtrl,
...
-
+void wxStyledTextCtrl::AutoCompShow(int lenEntered, const wxVector<wxCompletionItem>& list)
+{
+    m_autocompList.clear();
+    m_autocompList = list;
+    //Create correct format for Scintilla
+    wxString itemListString;
+    for(int i=0; i<m_autocompList.size(); i++)
+    {
+        wxCompletionItem cItem = m_autocompList[i];
```
Comment (vadz) [inline src/stc/stc.cpp]:
```suggestion
        const wxCompletionItem& cItem = m_autocompList[i];
```
no need to copy it needlessly.

File: src/stc/stc.cpp [hunk 9]
```diff
@@ -5062,7 +5062,19 @@ void wxStyledTextCtrl::AutoCompUseListCtrl(bool useListCtrl,
...
 }
 
-
+void wxStyledTextCtrl::AutoCompShow(int lenEntered, const wxVector<wxCompletionItem>& list)
+{
+    m_autocompList.clear();
+    m_autocompList = list;
+    //Create correct format for Scintilla
+    wxString itemListString;
+    for(int i=0; i<m_autocompList.size(); i++)
```
Comment (vadz) [inline src/stc/stc.cpp]:
```suggestion
    for ( int i = 0; i < m_autocompList.size(); i++ )
```
please try to use style consistent with the rest of wx

**Review by vadz — CHANGES_REQUESTED** [review]
Thanks, but we really need documentation for the new function, if only to explain to me how is this supposed to work.

An example in the sample would be helpful as well.

TIA!

File: .gitignore (hunk 1 above)
Comment (mtangoo) [reply .gitignore]: can you  explain where exactly is the global ignore? 
I thought this was the right place!

File: include/wx/stc/completionitem.h (hunk 3 above)
Comment (mtangoo) [reply include/wx/stc/completionitem.h]: It was not intended to be polymorphic for one reason: the only way I knew to alert the user when item is selected is attaching a wxClient object. 

wxSTC events don't have an option to set payload. How will I post it back selected object on event? Do you have an idea? If I can solve this then it will be polymorphic and client object stuff will be removed.

File: include/wx/stc/completionitem.h (hunk 4 above)
Comment (mtangoo) [reply include/wx/stc/completionitem.h]: good catch :)

File: include/wx/stc/stc.h (hunk 5 above)
Comment (mtangoo) [reply include/wx/stc/stc.h]: thank you!

File: include/wx/stc/stc.h (hunk 6 above)
Comment (mtangoo) [reply include/wx/stc/stc.h]: it was for my personal use and forgot to zap it off

File: src/stc/completionitem.cpp (hunk 7 above)
Comment (mtangoo) [reply src/stc/completionitem.cpp]: I don't get this well, sorry! can you help me understand?

File: src/stc/stc.cpp (hunk 8 above)
Comment (mtangoo) [reply src/stc/stc.cpp]: sure! Thanks

File: src/stc/stc.cpp (hunk 9 above)
Comment (mtangoo) [reply src/stc/stc.cpp]: Thank you

**Comment by mtangoo** [comment]
Hi @vadz,

> Thanks, but we really need documentation for the new function, if only to explain to me how is this supposed to work.

Of course I will do. I wanted to make sure I start right before writing everything only to find myself wrong. Now after all your review I feel confident to finish it up. I forgot on my points to add documentation. I will add it

>An example in the sample would be helpful as well.

Noted for action

File: .gitignore (hunk 1 above)
Comment (vadz) [reply .gitignore]: The value of `core.excludesfile` option, by default `~/.gitignore`.

File: src/stc/completionitem.cpp (hunk 7 above)
Comment (vadz) [reply src/stc/completionitem.cpp]: Just remove the new `.cpp` file and put the (little) code there is into the header.

File: .gitignore (hunk 1 above)
Comment (mtangoo) [reply .gitignore]: Ok noted, thank you!

**Comment by mtangoo** [comment]
@vadz I think am done as to what I planned to do. I have postponed to add The documentation Window because after this merge (if it ever happened to be merged) I wan't to work with @NewPagodi  to include that in his branch (and eventually PR). So If the PR is good enough, then you can merge it.
